File: frontend/src/services/aiService.js
import { Platform } from 'react-native';
import * as localDb from './localDbService';
import userService from './userService';

// Google Cloud AI API URL (백엔드와 동일)
const AI_API_URL = 'https://smartpot-api-551846265142.asia-northeast3.run.app/infer';

/**
 * AI 이미지 분석 서비스
 * - 식물 종류 판별 (mode="plant")
 * - 병충해 판별 (mode="disease")
 */

/**
 * 식물 종류 분석만 수행 (저장하지 않음)
 * @param {string} imageUri - 이미지 파일 경로 (file://)
 * @param {string} originalFileName - 원본 파일명 (선택, 없으면 자동 생성)
 * @returns {Promise<Object>} - AI 분석 결과 + plants DB 정보
 */
export const analyzeSpecies = async (imageUri, originalFileName = null) => {
  try {
    console.log('[analyzeSpecies] 시작:', { imageUri, originalFileName, platform: Platform.OS });

    // 파일명 결정
    let filename = originalFileName || imageUri.split('/').pop();
    const match = /\.(\w+)$/.exec(filename);
    const type = match ? `image/${match[1]}` : 'image/jpeg';

    if (!match) {
      const ext = type === 'image/png' ? 'png' : 'jpg';
      filename = `${filename}.${ext}`;
    }

    if (!filename.startsWith('plant_') && !filename.startsWith('leaf_')) {
      filename = `plant_${filename}`;
    }

    // FormData 생성
    const formData = new FormData();

    // 플랫폼별 처리
    if (Platform.OS === 'web') {
      if (imageUri.startsWith('blob:') || imageUri.startsWith('http')) {
        const blobResponse = await fetch(imageUri);
        const blob = await blobResponse.blob();
        const file = new File([blob], filename, { type: type });
        formData.append('file', file);
      } else {
        throw new Error('Web에서는 blob URL이 필요합니다');
      }
    } else {
      formData.append('file', {
        uri: imageUri,
        name: filename,
        type: type,
      });
    }

    console.log('[analyzeSpecies] Google Cloud AI API 호출 시작');

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 30000);

    try {
      // Google Cloud AI API 호출
      const response = await fetch(AI_API_URL, {
        method: 'POST',
        body: formData,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      console.log('[analyzeSpecies] AI API 응답 상태:', response.status);

      if (!response.ok) {
        const errorText = await response.text();
        console.error('[analyzeSpecies] AI API 오류:', errorText);
        throw new Error(`AI API 오류 (${response.status})`);
      }

      const aiData = await response.json();
      console.log('[analyzeSpecies] AI 분석 결과:', aiData);

      const predLabelEn = aiData.pred_label;
      const predLabelKo = aiData.pred_label_ko;

      // 로컬 DB에서 식물 정보 검색
      const plants = await localDb.searchPlants(predLabelKo);
      let plantInfo = null;

      if (plants && plants.length > 0) {
        const plant = plants[0];
        plantInfo = {
          plant_id: plant.id,
          ai_label_en: plant.ai_label_en,
          ai_label_ko: plant.ai_label_ko,
          wateringperiod: plant.wateringperiod || 7,
          ideallight: plant.ideallight,
          toleratedlight: plant.toleratedlight,
          watering: plant.watering,
          tempmin_celsius: plant.tempmin_celsius,
          tempmax_celsius: plant.tempmax_celsius,
        };
      }

      return {
        success: true,
        aiLabelEn: predLabelEn,
        aiLabelKo: predLabelKo,
        confidence: aiData.confidence,
        plantInfo: plantInfo,
      };
    } catch (fetchError) {
      clearTimeout(timeoutId);
      if (fetchError.name === 'AbortError') {
        throw new Error('요청 시간 초과 (30초)');
      }
      throw fetchError;
    }
  } catch (error) {
    console.error('[analyzeSpecies] 에러:', error);
    throw error;
  }
};

/**
 * 식물 종류 판별 (카메라로 찍은 식물 사진 분석)
 * @param {number} userId - 사용자 ID
 * @param {string} imageUri - 이미지 파일 경로 (file://)
 * @param {string} nickname - 식물 닉네임 (선택)
 * @param {string} originalFileName - 원본 파일명 (선택)
 * @returns {Promise<Object>} - AI 분석 결과 및 생성된 user_plant 정보
 */
export const identifySpecies = async (userId, imageUri, nickname = null, originalFileName = null) => {
  try {
    console.log('[identifySpecies] 시작:', { userId, imageUri, nickname, platform: Platform.OS });

    // 파일명 결정
    let filename = originalFileName || imageUri.split('/').pop();
    const match = /\.(\w+)$/.exec(filename);
    const type = match ? `image/${match[1]}` : 'image/jpeg';

    if (!match) {
      const ext = type === 'image/png' ? 'png' : 'jpg';
      filename = `${filename}.${ext}`;
    }

    if (!filename.startsWith('plant_') && !filename.startsWith('leaf_')) {
      filename = `plant_${filename}`;
    }

    // FormData 생성
    const formData = new FormData();

    // 플랫폼별 처리
    if (Platform.OS === 'web') {
      if (imageUri.startsWith('blob:') || imageUri.startsWith('http')) {
        const blobResponse = await fetch(imageUri);
        const blob = await blobResponse.blob();
        const file = new File([blob], filename, { type: type });
        formData.append('file', file);
      } else {
        throw new Error('Web에서는 blob URL이 필요합니다');
      }
    } else {
      formData.append('file', {
        uri: imageUri,
        name: filename,
        type: type,
      });
    }

    console.log('[identifySpecies] Google Cloud AI API 호출 시작');

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 30000);

    try {
      // Google Cloud AI API 직접 호출
      const response = await fetch(AI_API_URL, {
        method: 'POST',
        body: formData,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      console.log('[identifySpecies] AI API 응답 상태:', response.status);

      if (!response.ok) {
        const errorText = await response.text();
        console.error('[identifySpecies] AI API 오류:', errorText);
        throw new Error(`AI API 오류 (${response.status})`);
      }

      const aiData = await response.json();
      console.log('[identifySpecies] AI 분석 결과:', aiData);

      const predLabelEn = aiData.pred_label;
      const predLabelKo = aiData.pred_label_ko;

      // 로컬 DB에서 식물 검색
      const plants = await localDb.searchPlants(predLabelKo);
      let plantId = null;
      let wateringPeriod = 7;

      if (plants && plants.length > 0) {
        plantId = plants[0].id;
        wateringPeriod = plants[0].wateringperiod || 7;
      }

      // 사용자 식물 추가
      const userPlantId = await localDb.addUserPlant(
        userId,
        plantId,
        nickname || predLabelKo,
        imageUri,
        predLabelEn,
        predLabelKo,
        wateringPeriod
      );

      // 생성된 식물 조회
      const userPlants = await localDb.getUserPlants(userId);
      const createdPlant = userPlants.find(p => p.id === userPlantId);

      return {
        success: true,
        userPlant: createdPlant,
        aiResult: {
          species: predLabelEn,
          speciesKo: predLabelKo,
          confidence: aiData.confidence,
        },
      };
    } catch (fetchError) {
      clearTimeout(timeoutId);
      if (fetchError.name === 'AbortError') {
        throw new Error('요청 시간 초과 (30초)');
      }
      throw fetchError;
    }
  } catch (error) {
    console.error('[identifySpecies] 에러:', error);
    throw error;
  }
};

/**
 * 병충해 판별 (식물 잎사귀 사진 분석)
 * @param {number} userPlantId - 사용자 식물 ID
 * @param {string} imageUri - 이미지 파일 경로 (file://)
 * @param {string} filename - 파일명 (leaf_ 프리픽스 포함되어 전달됨)
 * @returns {Promise<Object>} - 병충해 진단 결과
 */
export const diagnoseDisease = async (userPlantId, imageUri, filename) => {
  try {
    console.log('[diagnoseDisease] 시작:', { userPlantId, imageUri, filename, platform: Platform.OS });

    // 파일명이 제공되지 않은 경우 기본값 사용
    if (!filename) {
      filename = `leaf_${Date.now()}.jpg`;
    }

    const match = /\.(\w+)$/.exec(filename);
    const type = match ? `image/${match[1]}` : 'image/jpeg';

    // FormData 생성
    const formData = new FormData();

    // 플랫폼별로 다르게 처리
    if (Platform.OS === 'web') {
      console.log('[diagnoseDisease] Web 플랫폼 처리');
      if (imageUri.startsWith('blob:') || imageUri.startsWith('http')) {
        const blobResponse = await fetch(imageUri);
        const blob = await blobResponse.blob();
        const file = new File([blob], filename, { type: type });
        formData.append('file', file);
      } else {
        throw new Error('Web에서는 blob URL이 필요합니다');
      }
    } else {
      console.log('[diagnoseDisease] React Native 플랫폼 처리');
      formData.append('file', {
        uri: imageUri,
        name: filename,
        type: type,
      });
    }

    console.log('[diagnoseDisease] Google Cloud AI API 호출 시작');

    // Google Cloud AI API 직접 호출 with timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => {
      console.log('[diagnoseDisease] 30초 타임아웃');
      controller.abort();
    }, 30000);

    try {
      const response = await fetch(AI_API_URL, {
        method: 'POST',
        body: formData,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      console.log('[diagnoseDisease] AI API 응답 상태:', response.status);

      if (!response.ok) {
        const errorText = await response.text();
        console.error('[diagnoseDisease] AI API 오류:', errorText);
        throw new Error(`AI API 오류 (${response.status})`);
      }

      const aiData = await response.json();
      console.log('[diagnoseDisease] AI 분석 결과:', aiData);

      const disease = aiData.pred_label_ko || aiData.pred_label;

      // 로컬 DB에 병충해 정보 저장
      await localDb.updateDisease(userPlantId, disease);

      return {
        success: true,
        userPlantId: userPlantId,
        disease: disease,
        aiResult: {
          disease: disease,
          confidence: aiData.confidence,
        },
      };
    } catch (fetchError) {
      clearTimeout(timeoutId);
      if (fetchError.name === 'AbortError') {
        console.error('[diagnoseDisease] 타임아웃 발생');
        throw new Error('요청 시간 초과 (30초). 네트워크 연결을 확인해주세요.');
      }
      throw fetchError;
    }
  } catch (error) {
    console.error('[diagnoseDisease] 최종 에러:', error);
    throw error;
  }
};

export default {
  analyzeSpecies,
  identifySpecies,
  diagnoseDisease,
};

File: backend/src/services/ai_service.py
"""
AI Service - 식물 종류 및 병충해 판별
Google Cloud AI API와 통신하여 이미지 분석
"""

import logging
import requests
from typing import Optional
from repositories import PlantRepository, UserPlantRepository

logger = logging.getLogger(__name__)


class AIService:
    # AI 이미지 분석 서비스

    AI_API_URL = "https://smartpot-api-551846265142.asia-northeast3.run.app/infer"

    def __init__(
        self,
        plant_repository: PlantRepository = None,
        user_plant_repository: UserPlantRepository = None,
    ):
        self.plant_repo = plant_repository or PlantRepository()
        self.user_plant_repo = user_plant_repository or UserPlantRepository()

    def analyze_image(self, image_file) -> dict:
        """
        AI API에 이미지를 전송하여 분석 결과 받기

        Args:
            image_file: 이미지 파일 객체 (Flask request.files['file'])

        Returns:
            dict: AI 분석 결과
            {
                "mode": "plant" 또는 "disease",
                "pred_label": "primula",
                "pred_label_ko": "앵초",
                "confidence": 0.63
            }
        """
        try:
            # 스트림을 처음으로 되돌림 (이미 읽혀진 경우 대비)
            image_file.stream.seek(0)

            # 파일 내용을 읽어서 전송 (스트림 문제 방지)
            file_content = image_file.read()

            # Google Cloud AI API에 맞는 형식으로 전송
            files = {"file": (image_file.filename, file_content, image_file.content_type)}

            logger.debug(f"AI API 요청 전송: {self.AI_API_URL}")
            logger.debug(f"파일명: {image_file.filename}, 크기: {len(file_content)} bytes")

            response = requests.post(self.AI_API_URL, files=files, timeout=30)

            logger.debug(f"AI API 응답 코드: {response.status_code}")
            logger.debug(f"AI API 응답 내용: {response.text[:200]}")

            if response.status_code == 200:
                result = response.json()
                return {
                    "success": True,
                    "mode": result.get("mode"),
                    "pred_label": result.get("pred_label"),
                    "pred_label_ko": result.get("pred_label_ko"),
                    "confidence": result.get("confidence"),
                }
            else:
                return {"success": False, "error": f"AI API error: {response.status_code}", "details": response.text}

        except requests.exceptions.Timeout:
            return {"success": False, "error": "AI API timeout"}
        except Exception as e:
            logger.error(f"AI 분석 중 예외 발생: {e}")
            return {"success": False, "error": str(e)}

    def analyze_species_only(self, image_file) -> dict:
        """
        식물 종류 분석만 수행 (저장하지 않음)
        사용자가 결과를 확인하고 직접 저장할 수 있도록 분석 결과만 반환

        Args:
            image_file: 이미지 파일 객체

        Returns:
            dict: AI 분석 결과 + plants DB 검색 결과
        """
        # AI 분석
        ai_result = self.analyze_image(image_file)

        if not ai_result.get("success"):
            return ai_result

        pred_label = ai_result.get("pred_label")
        pred_label_ko = ai_result.get("pred_label_ko")

        # plants DB에서 검색
        plants = self.plant_repo.search(pred_label_ko)  # 한글 라벨로 검색

        plant_info = None
        if plants:
            plant = plants[0]
            plant_info = {
                "plant_id": plant.id,
                "ai_label_en": plant.ai_label_en,
                "ai_label_ko": plant.ai_label_ko,
                "wateringperiod": plant.wateringperiod or 7,
                "ideallight": plant.ideallight,
                "toleratedlight": plant.toleratedlight,
                "watering": plant.watering,
                "tempmin_celsius": plant.tempmin_celsius,
                "tempmax_celsius": plant.tempmax_celsius,
            }

        return {
            "success": True,
            "ai_label_en": pred_label,
            "ai_label_ko": pred_label_ko,
            "confidence": ai_result.get("confidence"),
            "plant_info": plant_info,  # DB에서 찾은 식물 정보 (없으면 None)
        }

    def identify_species(self, user_id: int, image_file, nickname: str = None, image_path: str = None) -> dict:
        """
        식물 종류 판별 (mode="plant")
        pred_label로 plants DB 검색 후 user_plant 생성

        Args:
            user_id: 사용자 ID
            image_file: 이미지 파일 객체
            nickname: 식물 닉네임 (선택)
            image_path: 이미지 저장 경로 (선택)

        Returns:
            dict: 생성된 user_plant 정보
        """
        # AI 분석
        ai_result = self.analyze_image(image_file)

        if not ai_result.get("success"):
            return ai_result

        pred_label = ai_result.get("pred_label")
        pred_label_ko = ai_result.get("pred_label_ko")

        # plants DB에서 검색 (한글 라벨로)
        plants = self.plant_repo.search(pred_label_ko)

        if plants:
            # 찾은 식물 사용
            plant = plants[0]
            plant_id = plant.id
            wateringperiod = plant.wateringperiod or 7
        else:
            # 못 찾으면 plant_id=None 사용
            plant_id = None
            wateringperiod = 7

        # user_plant 생성
        user_plant_id = self.user_plant_repo.save(
            user_id=user_id,
            plant_id=plant_id,
            nickname=nickname or pred_label_ko,
            image=image_path,
            ai_label_en=pred_label,
            ai_label_ko=pred_label_ko,
            wateringperiod=wateringperiod,
        )

        # 생성된 정보 조회
        user_plants = self.user_plant_repo.find_by_user(user_id)
        created_plant = next((p for p in user_plants if p["id"] == user_plant_id), None)

        return {
            "success": True,
            "mode": "plant",
            "user_plant": created_plant,
            "ai_result": ai_result,
        }

    def diagnose_disease(self, user_plant_id: int, image_file) -> dict:
        """
        병충해 판별 (mode="disease")
        pred_label_ko를 user_plant의 disease 필드에 저장

        Args:
            user_plant_id: 사용자 식물 ID
            image_file: 이미지 파일 객체

        Returns:
            dict: 병충해 진단 결과
        """
        # AI 분석
        ai_result = self.analyze_image(image_file)

        if not ai_result.get("success"):
            return ai_result

        pred_label_ko = ai_result.get("pred_label_ko")

        # user_plant의 disease 필드 업데이트
        success = self.user_plant_repo.update(user_plant_id=user_plant_id, disease=pred_label_ko)

        if not success:
            return {"success": False, "error": "Failed to update disease information"}

        return {
            "success": True,
            "mode": "disease",
            "user_plant_id": user_plant_id,
            "disease": pred_label_ko,
            "ai_result": ai_result,
        }
